### DISCUSSION TITLE
performance of realtime events

### DISCUSSION THREAD

**Discussion by godmar** [pocketbase/pocketbase#5052] [Q&A]
I was wondering how pocketbase implements its realtime updates. Looking at [realtime.go](https://github.com/pocketbase/pocketbase/blob/master/apis/realtime.go#L296-L308), it appears that on every record creation, update, and deletion an event fires that calls `broadcastRecord`.  In that function, the code [iterates over all connected clients](https://github.com/pocketbase/pocketbase/blob/master/apis/realtime.go#L394), and for each client, [checks all six possible subscription modes](https://github.com/pocketbase/pocketbase/blob/master/apis/realtime.go#L399) to see whether the client has subscribed to them. If so, a copy of the record is made and then a check is performed if the [client can access this record](https://github.com/pocketbase/pocketbase/blob/master/apis/realtime.go#L420). If the subscription doesn't use a filter, the access rules is checked; otherwise a query is built and executed.

I'm a bit concerned about the performance/complexity of this approach. 

First, why aren't the clients indexed by their subscription?  This way, instead of iterating over all connected clients - which could be many - it would iterate only over the clients that are subscribed to one of the events on that collection.

Second, in the case where the subscription is for a collection and where it uses a filter, it appears that the `canAccessRecord` method runs a [database query](https://github.com/pocketbase/pocketbase/blob/970b00011f45f9da40a6b212c307ad0af3a3aaa9/apis/realtime.go#L554) to find out whether the value matches the filter or not.  This would mean that in addition to the checks and copies performed so far, every record update would run $n$ queries where $n$ is the number of subscribers to this table. In [issue #1657 ](https://github.com/pocketbase/pocketbase/issues/1657#issuecomment-1407734794) Gani also voiced some unspecified concern about the performance of filters.

Here is the potential use case that led me to this thinking about this concern: suppose you have a simple 3-field highscore table `(gameid, playerid, score)`. These represent an M:N relationship between games and players. You want realtime updates whenever a score changes for any `playerid` within a given `gameid` - therefore, you need to subscribe to the entire table and apply a filter on the `gameid` under the current design.  This would mean that if *any* gameid has an update, all subscribed clients would run a db query. 

Moreover, if a record is updated that belongs to the game of interest, the subscribed client will only obtain the changed record. Ordinarily, they would then run a request (say via a View collection) to get the highscore board for their game by projecting to their gameid (unless they maintain a copy of all scores on the client side, which is awkward). 

I'm concerned that this will not scale even within the small to medium dimensions for which pocketbase is designed.  (I probably wouldn't even try it and use a polling-based approach instead.)

By comparison, other pubsub systems I'm familiar with (say [Redis Pub/Sub](https://redis.io/docs/latest/develop/interact/pubsub/)) typically use a concept of channels which allows the server to efficiently identify who should receive a message that is published. 

Of course, in pocketbase's case, you'll want to tie this to the database which forms the basis of all this, but I'm wondering if thinking about a concept of channels wouldn't make sense here, too. For instance, creates/edits/deletes could be associated with channels that they update, and subscriptions would not (only) be to records/collections, but to channels instead (or in addition to the existing ideas).  The channel name could be tied to the collection name or to the values of individual fields, perhaps.  In my use case, every score update would trigger a pub to the channel associated with the gameid, and clients would subscribe to the channel of the game they're currently playing, rather than the entire table.

I should say that I really like pocketbase so far and so I'd love to be able to use its realtime features as well.  Any comments are appreciated.

**Comment by ganigeorgiev** [comment]
The short answer is, because it is simple, easy to reason and hasn't cause an issue so far (at least for me).

About your first point - the subscriptions could be practically anything (including custom user events). Additionally one client could have multiple subscriptions which would complicate the unsubscribe/disconnect flow ever further down the road. Iterating over 100K+ entries shouldn't be that much of an issue (note that we also  exit early if the client doesn't have subscriptions matching the topic).

For your second point - during the implementation I found that the filter access checks weren't an issue (_they are essentially a query against a single primary key with some additional `WHERE` constraints_). Additionally note that [Many Small Queries Are Efficient In SQLite](https://www.sqlite.org/np1queryprob.html).

Suggestions for improvements are welcomed but I'm not willing to invest time complicating the implementation at the moment unless there is an actual issue.

With that said, I'll accept a PR if the suggested implementation performs significantly better.

**Reply by godmar** [reply]
I think it may be difficult to produce an issue unless you're doing testing or deployment at scale. In the [FAQ](https://pocketbase.io/faq/) you mention that you're targeting 10000+ persistent realtime connections.

Regarding the first observation: in a typical use case, I would expect all connections to subscribe to the same realtime events. Even in the case where all subscriptions are for single records, this would execute your logic 10000 x 6 times for every single update to a table (at least [this test](https://github.com/pocketbase/pocketbase/blob/970b00011f45f9da40a6b212c307ad0af3a3aaa9/apis/realtime.go#L400)). My intuition is that this could lead to noticeable performance degradation especially during bursty periods, such as when many users perform simultaneous updates.

You are correct that this would complicate the logic of managing subscriptions, though I don't follow what you mean by "custom user events" - I saw only the [6 possible subscription paths](https://github.com/pocketbase/pocketbase/blob/970b00011f45f9da40a6b212c307ad0af3a3aaa9/apis/realtime.go#L382-L390) in the code.

**Reply by ganigeorgiev** [reply]
> ... though I don't follow what you mean by "custom user events" ...

`broadcastRecord` is just one way to send a realtime event but users can register their own custom events and notify the specific `app.SubscriptionsBroker()` clients (_to listen to those events from the SDKs: `pb.realtime.subscribe("topic", (e) => { ... })`_).

Again I'm eager to change the implementation if there is a real improvement without complicating too much the existing APIs but so far I haven't stumbled on any major problems and I'm not planning investing time on this unless there is an actual issue.

If the clients loop is the main cause of your concerns I guess one simple optimization could be to split it and execute it concurrently, for example X clients per goroutine. I'll consider this with the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355).

I think the performance for now is "acceptable" for the type of applications PocketBase is targeting. The overall DX and stabilizing the APIs remains with higher priority. Detailed profiling and more complex performance tuning will be done after all other major tasks are completed.

**Reply by godmar** [reply]
Thank you for this pointer - perhaps these will address my use case better and avoid having to add a filtering-based realtime subscription on the entire collection.

In the js-sdk documentation, I found the section on [Realtime Services](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#realtimeservice) that describe the `realtime` API you mention.  However, it's not clear how such custom events are produced.
I'm guessing that this is done by adding a hook in `pb_hooks` server side, or is there a way from the client API to broadcast a custom event?

If it's done from a custom event, how exactly does it work?  [`subscriptions.Broker`](https://pocketbase.io/jsvm/interfaces/subscriptions.Broker.html) has a `.clients()` method; and the [`Client`](https://pocketbase.io/jsvm/interfaces/subscriptions.Client.html#hasSubscription) class has a `hasSubscription` and `send` method.  Would I need to write a JS loop kind of like:
```js
for (let client of broker.clients())
   if (client.hasSubscription("mytopic"))
       client.send(message)
```
? or is there an API I'm missing?

**Reply by ganigeorgiev** [reply]
> I'm guessing that this is done by adding a hook in pb_hooks server side, or is there a way from the client API to broadcast a custom event?

Custom realtime events can be triggered only via the app event hooks (with [Go](https://pocketbase.io/docs/go-event-hooks/) or the [JSVM `pb_hooks`](https://pocketbase.io/docs/js-event-hooks/)).

> Would I need to write a JS loop kind of like ...

Yes, we currently don't have a generic "broadcast" helper.

In any case this is getting too off topic. If you have further questions please elaborate more clearly on your use case to understand the bigger picture and the actual problem but I'm not willing to invest more time on some imaginary scenarios.

**Reply by ganigeorgiev** [reply]
_Just as a side-note in case you decide to use the JSVM for sending custom events - keep in mind that there is no concurrent execution within one JSVM handler so if you are planning to send custom events only because of some eventual performance gains you'll be better using Go. But my recommendation is to simply use the existing APIs until an actual problem arise._

**Comment by godmar** [comment]
Thank you for the pointer.  I don't feel that this is getting off topic, it's very helpful for me and perhaps for others reading this. Thank you.

From my perspective, there are 3 issues:

1. Whether it's worth optimizing the realtime implementation such that clients are indexed by their subscription. This would add some code and complexity to the registration and cleanup, but would optimize the matching of events to clients.

2. How to deal with use cases where a client would listen to realtime events on a filtered collection (as opposed to a single record) that may be expensive. I'm in particular thinking of cases where say you have 10,000 participants split into 500 active groups of 20 entries in a M:N relationship table and where events that update any of the 10,000 records must be broadcast to the entire group of 20 whenever an entry belonging to a group changes. This is an independent issue from 1. as the performance concern would apply even if 1. were implemented.

3. Closely related to 2., I had the question whether a general purpose realtime event broadcast notification facility would make sense - but you helpfully pointed out that Pocketbase already implements one! Indeed, with your hint, I believe I am able to cover my use case with a hook like so:

```js
onModelAfterUpdate((e) => {
    const game = e.model.get('contest');
    const broker = $app.subscriptionsBroker();
    const clients = broker.clients();
    for (let [clientId, client] of Object.entries(clients)) {
        if (client.hasSubscription(game)) {
            const msg = new SubscriptionMessage({
                name: game,  // must match the topic name the client subscribes to
                data: JSON.stringify(e.model)
            });
            client.send(msg);
        }   
    }
}, "results")
```
which I was able to successfully test. (**)

Here, clients can subscribe to the "`contest`" subgroup of results they're interested in, and updates to any `result` entries will be broadcast to them. It would benefit from the optimization in 1.

This works, although understandably uses a lot of the wide and detailed JSVM extensibility surface you provide. On the other hand, doing without hooks would be even better, if for no other reason than that these interfaces are unlikely to be stable in the future (and they probably shouldn't be).

It got me thinking though whether providing a general-purpose pub/sub facility that's not necessarily tied to the database wouldn't make sense. As of now, there doesn't appear to be a `pb.realtime.publish('TOPIC', event)` call that would send a client-triggered event to all subscribers.
Should there be one?  It would probably require adding some kind of access control, perhaps by some kind of integration with the users auth collection, and/or support for a set of rules that would regulate whether clients could send messages and if so, which ones. Perhaps also some namespaces. Or does this exist already?  I would argue in favor of supporting this, if for no other reason than the options so far (which are per-record subscriptions and per-collection subscriptions) may not cover all use cases (see my example.) (*)

Thank you for reading.

(*) I believe I could have used per-record subscriptions, which I believe would benefit from the optimization in 1., but in my data schema I believe that this would have come at the expense of an additional API call (e.g., trigger some kind of update in the `game` table for the specific game record.)

(**) BTW, I'm not sure I need to be too concerned about the loss of parallelism due to the single-threaded nature of the JSVM, given that the entire app is supposed to run on a cheap VM with only 1 or 2 VCPUs purchased, and it's a send() rather than recv() which normally shouldn't block.
